### PR TITLE
Generate reflection declaration from MP OpenAPI schema definition

### LINKED ISSUE #745: JAX-RS - Generate reflection declaration from MP OpenAPI schema definition
It is possible to define the entity types via the MP OpenAPI annotations. Look for an example here:
https://openliberty.io/guides/microprofile-openapi.html#augmenting-the-existing-jax-rs-annotations-with-openapi-annotations

These entity types should be registered for reflection as they will probably be serialized at some point.

AFAICS, in `@Schema`, there is `implementation` but we should probably also take into account `oneOf` and friends.

See what we already do here: https://github.com/quarkusio/quarkus/blob/master/extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyScanningProcessor.java#L517 .

### PR DESCRIPTION
**Pull request by machi1990**
Fixes #745.

The main idea is to scan for SCHEMA annotation and mark them as ReflectiveHierarchy
-Scan APIResponse annotation for schema declaration and mark the results ( `implementation , not, oneOf, anyOf, allOf values`) for reflection.
-Scan APIResponses annotation and do as above.

### CODE REVIEW

**Comment by dmlloyd** [comment]
Could you please provide a descriptive PR summary?  It's OK to have "Fixes..." in the body but the title should say what the PR is actually for.

**Comment by machi1990** [comment]
@dmlloyd I have renamed the PR. Should the same be done for the commit message?

**Comment by dmlloyd** [comment]
> @dmlloyd I have renamed the PR. Should the same be done for the commit message?

Yes please.

**Review by gsmet — CHANGES_REQUESTED** [review]
@machi1990 I added another round of comments. We're nearly there!

File: extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyScanningProcessor.java
Comment (machi1990) [reply extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyScanningProcessor.java]: Seems great. Something like `OPENAPI_SCHEMA_NOT` ?

File: integration-tests/main/src/main/java/io/quarkus/example/rest/TestResource.java
Comment (machi1990) [reply integration-tests/main/src/main/java/io/quarkus/example/rest/TestResource.java]: Good catch.

File: integration-tests/main/src/main/java/io/quarkus/example/rest/TestResource.java
Comment (machi1990) [reply integration-tests/main/src/main/java/io/quarkus/example/rest/TestResource.java]: No need. I wanted to test @Schema scanning when a schema class is the annotation target. Can be removed.

File: extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyScanningProcessor.java
Comment (gsmet) [reply extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyScanningProcessor.java]: Yes, exactly.

File: integration-tests/main/src/main/java/io/quarkus/example/rest/TestResource.java
Comment (gsmet) [reply integration-tests/main/src/main/java/io/quarkus/example/rest/TestResource.java]: OK, so I think testing that is required but... you should do that in a specific test. Otherwise if one part is working but not the other, your test will be successful anyway.

You need one test per feature in this case.

**Comment by gsmet** [comment]
@machi1990 I let you finish this? Looks like we just need some minor adjustments to the constants and the tests, now.

**Comment by machi1990** [comment]
> @machi1990 I let you finish this? Looks like we just need some minor adjustments to the constants and the tests, now.
 
@gsmet looks like we are good. I do not understand why the build is failing on the CI though.

**Comment by machi1990** [comment]
> > @machi1990 I let you finish this? Looks like we just need some minor adjustments to the constants and the tests, now.
> 
> @gsmet looks like we are good. I do not understand why the build is failing on the CI though.

**Comment by machi1990** [comment]
@gsmet I have finished integrating the requested changes and rebased the branch with master. This is ready for another round of review.

**Comment by gsmet** [comment]
Thanks, it's on my TODO list for today.

**Comment by gsmet** [comment]
@machi1990 I just force pushed a rebase to hopefully get CI working correctly. Review planned for later today.

File: integration-tests/main/src/test/java/io/quarkus/example/test/JaxRSTestCase.java
Comment (gsmet) [inline integration-tests/main/src/test/java/io/quarkus/example/test/JaxRSTestCase.java]: You're not testing the `"/openapi/responses/{version}"` path.

Otherwise, it looks good.

Could you rebase and take care of it?

**Review by gsmet — CHANGES_REQUESTED** [review]

File: integration-tests/main/src/test/java/io/quarkus/example/test/JaxRSTestCase.java
Comment (machi1990) [reply integration-tests/main/src/test/java/io/quarkus/example/test/JaxRSTestCase.java]: Great. I have rebased the master and force pushed to the branch.

**Review by gsmet — APPROVED** [review]
Looks good to me, thanks!

I'll wait for CI to go green and merge.

**Comment by machi1990** [comment]
Thanks to you too. I can contribute to something else if need be.

**Comment by kenfinnigan** [comment]
Coming late to this, but wondering whether it makes sense for this to be done as part of the `quarkus-smallrye-openapi` extension?

Mostly because if you're using MP OpenAPI annotations, without adding that dependency yourself you're only going to have that if the `quarkus-smallrye-openapi` extension is installed.

**Comment by gsmet** [comment]
@kenfinnigan Aaaah... sorry, just saw your message when I clicked the Merge button as I was waiting for this to go green.

So the rationale was to try to centralize the serialization reflection registration so that we have a global view of what's going on in the RESTEasy extension.

That's something we can discuss for sure, we can move the code if needed.

**Comment by kenfinnigan** [comment]
Just seems odd for RESTEasy extension to need to know about OpenAPI, and potentially lots of other things

**Comment by machi1990** [comment]
Oops, I too moved a bit fast and deleted my local branch. 

I think @kenfinnigan is making a interesting point, and thanks to @gsmet for giving an element of response to the case. I took this approach for the reason stated above. Should we decide to move the code to `quarkus-smallrye-openapi` extension, I'll be up for the task.

**Comment by gsmet** [comment]
@machi1990 Yeah, maybe let's do that.

It shouldn't be too much of a hassle: you should move the code in the processor (and use the annotation classes as they will be in the classpath now). The tests can stay as they are.

**Pull request merged**
